src/Features/CSharp/Portable/AddRequiredParentheses/CSharpAddRequiredParenthesesDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,80 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Immutable;
+using System.Composition;
+using System.Diagnostics;
+using Microsoft.CodeAnalysis.AddRequiredParentheses;
+using Microsoft.CodeAnalysis.CSharp.Extensions;
+using Microsoft.CodeAnalysis.CSharp.RemoveUnnecessaryParentheses;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.RemoveUnnecessaryParentheses;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp.AddRequiredParentheses
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp), Shared]
+    internal class CSharpAddRequiredParenthesesDiagnosticAnalyzer :
+        AbstractAddRequiredParenthesesDiagnosticAnalyzer<
+            ExpressionSyntax, ExpressionSyntax, SyntaxKind>
+    {
+        private static readonly ImmutableArray<SyntaxKind> s_kinds = ImmutableArray.Create(
+            SyntaxKind.AddExpression,
+            SyntaxKind.SubtractExpression,
+            SyntaxKind.MultiplyExpression,
+            SyntaxKind.DivideExpression,
+            SyntaxKind.ModuloExpression,
+            SyntaxKind.LeftShiftExpression,
+            SyntaxKind.RightShiftExpression,
+            SyntaxKind.LogicalOrExpression,
+            SyntaxKind.LogicalAndExpression,
+            SyntaxKind.BitwiseOrExpression,
+            SyntaxKind.BitwiseAndExpression,
+            SyntaxKind.ExclusiveOrExpression,
+            SyntaxKind.EqualsExpression,
+            SyntaxKind.NotEqualsExpression,
+            SyntaxKind.LessThanExpression,
+            SyntaxKind.LessThanOrEqualExpression,
+            SyntaxKind.GreaterThanExpression,
+            SyntaxKind.GreaterThanOrEqualExpression,
+            SyntaxKind.IsExpression,
+            SyntaxKind.AsExpression,
+            SyntaxKind.CoalesceExpression,
+            SyntaxKind.IsPatternExpression);
+
+        protected override ImmutableArray<SyntaxKind> GetSyntaxNodeKinds()
+            => s_kinds;
+
+        protected override int GetPrecedence(ExpressionSyntax binaryLike)
+            => (int)binaryLike.GetOperatorPrecedence();
+
+        protected override PrecedenceKind GetPrecedenceKind(ExpressionSyntax binaryLike)
+            => CSharpRemoveUnnecessaryParenthesesDiagnosticAnalyzer.GetPrecedenceKind(binaryLike);
+
+        protected override bool IsBinaryLike(ExpressionSyntax node)
+            => node is BinaryExpressionSyntax ||
+               node is IsPatternExpressionSyntax isPattern && isPattern.Pattern is ConstantPatternSyntax;
+
+        protected override (ExpressionSyntax, SyntaxToken, ExpressionSyntax) GetPartsOfBinaryLike(ExpressionSyntax binaryLikeOpt)
```

Comment:
❕ The argument is not optional

:bulb: Named tuple elements

Reply:
renamed.  In this case i'm ok with non-named.  i think it's very clear from context.